src/file_op.rs

```diff
@@ -4,50 +4,115 @@ use std::io::ErrorKind;
 use tokio::io::AsyncWriteExt;
 
 use crate::{
     config::MiniserveConfig, errors::RuntimeError, file_utils::contains_symlink,
     file_utils::sanitize_path,
 };
 
+enum FileHash {
+    SHA256(String),
+}
+
+impl FileHash {
+    pub fn get_hasher(&self) -> impl Digest {
+        match self {
+            Self::SHA256(_) => Sha256::new(),
+        }
+    }
+}
+
 /// Saves file data from a multipart form field (`field`) to `file_path`, optionally overwriting
 /// existing file.
 ///
 /// Returns total bytes written to file.
 async fn save_file(
-    field: actix_multipart::Field,
+    field: &mut actix_multipart::Field,
     file_path: PathBuf,
     overwrite_files: bool,
+    file_hash: Option<&FileHash>,
 ) -> Result<u64, RuntimeError> {
     if !overwrite_files && file_path.exists() {
         return Err(RuntimeError::DuplicateFileError);
     }
 
-    let file = match File::create(&file_path).await {
-        Err(err) if err.kind() == ErrorKind::PermissionDenied => Err(
+    let named_temp_file = match tokio::task::spawn_blocking(|| NamedTempFile::new()).await {
+        Err(err) => Err(RuntimeError::MultipartError(format!(
+            "Failed to complete spawned task to create named temp file. {}",
+            err
+        ))),
+        Ok(Err(err)) if err.kind() == ErrorKind::PermissionDenied => Err(
             RuntimeError::InsufficientPermissionsError(file_path.display().to_string()),
         ),
-        Err(err) => Err(RuntimeError::IoError(
-            format!("Failed to create {}", file_path.display()),
+        Ok(Err(err)) => Err(RuntimeError::IoError(
+            format!("Failed to create temporary file {}", file_path.display()),
             err,
         )),
-        Ok(v) => Ok(v),
+        Ok(Ok(file)) => Ok(file),
     }?;
 
-    let (_, written_len) = field
-        .map_err(|x| RuntimeError::MultipartError(x.to_string()))
-        .try_fold((file, 0u64), |(mut file, written_len), bytes| async move {
-            file.write_all(bytes.as_ref())
-                .map_err(|e| RuntimeError::IoError("Failed to write to file".to_string(), e))
-                .await?;
-            Ok((file, written_len + bytes.len() as u64))
-        })
-        .await?;
+    let (file, temp_path) = named_temp_file.keep().map_err(|err| {
+        RuntimeError::IoError("Failed to keep temporary file".into(), err.error.into())
+    })?;
+    let mut temp_file = tokio::fs::File::from_std(file);
+
+    let mut written_len = 0;
+    let mut hasher = file_hash.as_ref().map(|h| h.get_hasher());
+    let mut error: Option<RuntimeError> = None;
+
+    while let Some(Ok(bytes)) = field.next().await {
+        if let Some(hasher) = hasher.as_mut() {
+            hasher.update(&bytes)
+        }
+        if let Err(e) = temp_file.write_all(&bytes).await {
+            error = Some(RuntimeError::IoError(
+                "Failed to write to file".to_string(),
+                e,
+            ));
+            break;
+        }
```

Comment:
Why do we even store the error here? Couldn't we just return `drop(temp_file);` and then `return` the error?

Reply:
We could do that, but we would be duplicating code because we'd need to call `drop(temp_file)` and `tokio::fs::remove_file(...)` anyways. I also noticed that I should have been calling `flush` on the file as well which will also update the existing error if it doesn't exist.

You'll be able to see my changes soon, but I don't think I support this suggestion.